abi.html

```diff
@@ -662,40 +693,110 @@ <h3><a href="#pod"> 2.2 POD Data Types </a></h3>
 <a name="member-pointers"></a>
 <h3><a href="#member-pointers"> 2.3 Member Pointers </a></h3>
 
+<a name="data-member-pointers"></a>
+<h4><a href="#data-member-pointers"> 2.3.1 Data Member Pointers </a></h4>
+
 <p>
-A pointer to data member is an offset from the base
-address of the class object containing it,
-represented as a <code>ptrdiff_t</code>.
-It has the size and alignment attributes of a <code>ptrdiff_t</code>.
-A NULL pointer is represented as -1.
+The basic ABI properties of data member pointer types are those
+of <code>ptrdiff_t</code>.
 
 <p>
-A pointer to member function is a pair <ptr, adj> as follows:
+A data member pointer is represented as the data member's offset in bytes
+from the address point of an object of the base type, as a
+<code>ptrdiff_t</code>.
 
-<dl>
 <p>
-<dt> <code>ptr</code>:
-<dd> For a non-virtual function, this field is a simple function pointer.
-     (Under current base Itanium psABI conventions,
-     that is a pointer to a GP/function address pair.)
-     For a virtual function,
-     it is 1 plus the virtual table offset (in bytes) of the function,
-     represented as a <code>ptrdiff_t</code>.
-     The value zero represents a NULL pointer,
-     independent of the adjustment field value below.
+A null data member pointer is represented as an offset of <code>-1</code>.
+Unfortunately, it is possible to generate an data member pointer with an
+offset of <code>-1</code> using explicit derived-to-base conversions.
+If this is done, implementations following this ABI may misbehave.
+<span class="future-abi">Recommendation for new platforms: consider using
+<code>PTRDIFF_MIN</code> as the null data member pointer instead.</span>
```

Comment:
I'm not sure whether we should be recommending this (and nothing else). Pointer-to-data-member casts generally need a branch or conditional move on nullness; is `PTRDIFF_MIN` easy enough to detect on most targets? (On x86, I suppose `cmp a 1; cmovno adjusted_value a` works.) If we're suggesting changing the representation anyway, I also think we should suggest a change that makes zero-initialization cheaper.

In particular, I think I would want people bringing up new targets to also consider storing the value left-shifted one place, using lowest bit == 0 for null pointers and lowest bit == 1 for non-null pointers. Pros: zero-initialization is all zeroes. Casts can be performed by unconditional adds. Null checks are fast. Con: require a (signed) right shift on every use.

Reply:
Yeah, good point.  Made a different recommendation.